File: data_converter.py
"""
JSON -> minimal YAML-ish flattening.

JSON spends an enormous share of its tokens on pure syntax: ``{``, ``}``,
``[``, ``]``, ``"`` around every key *and* every string value, and ``,``
between every pair. On nested config blobs that syntax is frequently 30-50%
of the token bill while carrying zero information.

``json_to_minimal_yaml`` emits a bracket-free, quote-free dotted-path form::

    service.name api-gateway
    service.ports.0 8080
    service.debug true

The transform is **reversible** (see :func:`minimal_yaml_to_json`), which is
what makes it lossless rather than merely lossy-but-short. Values that would
become ambiguous when unquoted (empty strings, leading/trailing whitespace,
strings that look like numbers/bools/null) are quoted so the round-trip is
exact.
"""

from __future__ import annotations

import json
import re
from typing import Any, Dict, List, Mapping, Sequence, Tuple

__all__ = [
    "json_to_minimal_yaml",
    "minimal_yaml_to_json",
]

# A bare string is safe only if it cannot be mistaken for another JSON scalar.
_LOOKS_NUMERIC = re.compile(r"^[+-]?(?:\d+\.?\d*|\.\d+)(?:[eE][+-]?\d+)?$")
_RESERVED = {"true", "false", "null", "none", "~", "-", ""}


def _needs_quoting(text: str) -> bool:
    if text != text.strip() or text == "":
        return True
    if text.lower() in _RESERVED:
        return True
    if _LOOKS_NUMERIC.match(text):
        return True
    if "\n" in text or "\r" in text or "\t" in text:
        return True
    if text[0] in "\"'#[{&*!|>%@`":
        return True
    return False


def _encode_scalar(value: Any) -> str:
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, (int, float)):
        return repr(value) if isinstance(value, float) else str(value)
    text = str(value)
    if _needs_quoting(text):
        return json.dumps(text, ensure_ascii=False)
    return text


def _decode_scalar(token: str) -> Any:
    if token.startswith('"'):
        return json.loads(token)
    if token == "null":
        return None
    if token == "true":
        return True
    if token == "false":
        return False
    if _LOOKS_NUMERIC.match(token):
        try:
            return int(token)
        except ValueError:
            return float(token)
    return token


def _escape_key(key: str) -> str:
    """Dots inside a key would collide with the path separator."""
    return key.replace("\\", "\\\\").replace(".", "\\.")


def _unescape_key(key: str) -> str:
    return key.replace("\\.", ".").replace("\\\\", "\\")


def _split_path(path: str) -> List[str]:
    parts: List[str] = []
    buf: List[str] = []
    i = 0
    while i < len(path):
        ch = path[i]
        if ch == "\\" and i + 1 < len(path):
            buf.append(path[i : i + 2])
            i += 2
            continue
        if ch == ".":
            parts.append("".join(buf))
            buf = []
            i += 1
            continue
        buf.append(ch)
        i += 1
    parts.append("".join(buf))
    return [_unescape_key(p) for p in parts]


def _flatten(node: Any, prefix: str, sink: List[Tuple[str, str]]) -> None:
    if isinstance(node, Mapping):
        if not node:
            sink.append((prefix, "{}"))
            return
        for key, value in node.items():
            child = _escape_key(str(key))
            _flatten(value, f"{prefix}.{child}" if prefix else child, sink)
        return

    if isinstance(node, Sequence) and not isinstance(node, (str, bytes, bytearray)):
        if not node:
            sink.append((prefix, "[]"))
            return
        # Arrays of plain scalars collapse onto one line: ports 80, 443, 8080
        if all(not isinstance(v, (Mapping, list, tuple)) for v in node):
            joined = ", ".join(_encode_scalar(v) for v in node)
            sink.append((prefix, joined))
            return
        for index, value in enumerate(node):
            _flatten(value, f"{prefix}.{index}" if prefix else str(index), sink)
        return

    sink.append((prefix, _encode_scalar(node)))


def json_to_minimal_yaml(data_obj: Any, *, root_key: str = "") -> str:
    """Flatten a nested dict/list into bracket-free, quote-free text.

    Accepts a Python object or a JSON string. Output is one ``path value``
    pair per line, dot-separated, with list indices as path segments.
    """
    if isinstance(data_obj, (str, bytes, bytearray)):
        data_obj = json.loads(data_obj)

    sink: List[Tuple[str, str]] = []
    _flatten(data_obj, root_key, sink)
    return "\n".join(f"{path} {value}".rstrip() for path, value in sink)


def minimal_yaml_to_json(text: str) -> Any:
    """Inverse of :func:`json_to_minimal_yaml` (proves the transform is lossless).

    Note: a scalar array is restored as a list; a single-element scalar array is
    indistinguishable from a bare scalar, which is the one documented ambiguity.
    """
    root: Dict[str, Any] = {}

    for line in text.splitlines():
        if not line.strip():
            continue
        path, _, raw_value = line.partition(" ")
        raw_value = raw_value.strip()
        parts = _split_path(path)

        if raw_value == "{}":
            value: Any = {}
        elif raw_value == "[]":
            value = []
        elif "," in raw_value and not raw_value.startswith('"'):
            value = [_decode_scalar(v.strip()) for v in _split_top_level(raw_value)]
        else:
            value = _decode_scalar(raw_value)

        cursor: Any = root
        for i, part in enumerate(parts):
            last = i == len(parts) - 1
            if last:
                cursor[part] = value
            else:
                cursor = cursor.setdefault(part, {})

    return _relist(root)


def _split_top_level(text: str) -> List[str]:
    """Split on commas that are not inside a quoted string."""
    out: List[str] = []
    buf: List[str] = []
    in_str = False
    i = 0
    while i < len(text):
        ch = text[i]
        if in_str:
            if ch == "\\":
                buf.append(text[i : i + 2])
                i += 2
                continue
            if ch == '"':
                in_str = False
            buf.append(ch)
        elif ch == '"':
            in_str = True
            buf.append(ch)
        elif ch == ",":
            out.append("".join(buf))
            buf = []
        else:
            buf.append(ch)
        i += 1
    out.append("".join(buf))
    return out


def _relist(node: Any) -> Any:
    """Turn {'0': .., '1': ..} maps back into lists, depth-first."""
    if not isinstance(node, dict):
        return node
    converted = {k: _relist(v) for k, v in node.items()}
    if converted and all(k.isdigit() for k in converted):
        expected = [str(i) for i in range(len(converted))]
        if sorted(converted.keys(), key=int) == expected:
            return [converted[k] for k in expected]
    return converted

File: rust/src/data_converter.rs
//! JSON -> minimal, bracket-free, quote-free text (and back again).

use once_cell::sync::Lazy;
use regex::Regex;
use serde_json::{Map, Value};

static LOOKS_NUMERIC: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^[+-]?(?:\d+\.?\d*|\.\d+)(?:[eE][+-]?\d+)?$").expect("valid numeric regex"));

const RESERVED: [&str; 6] = ["true", "false", "null", "none", "~", "-"];

fn needs_quoting(text: &str) -> bool {
    if text.is_empty() || text.trim() != text {
        return true;
    }
    let lower = text.to_ascii_lowercase();
    if RESERVED.contains(&lower.as_str()) {
        return true;
    }
    if LOOKS_NUMERIC.is_match(text) {
        return true;
    }
    if text.contains('\n') || text.contains('\r') || text.contains('\t') {
        return true;
    }
    matches!(
        text.chars().next(),
        Some('"') | Some('\'') | Some('#') | Some('[') | Some('{') | Some('&')
            | Some('*') | Some('!') | Some('|') | Some('>') | Some('%') | Some('@') | Some('`')
    )
}

fn encode_scalar(v: &Value) -> String {
    match v {
        Value::Null => "null".to_string(),
        Value::Bool(b) => b.to_string(),
        Value::Number(n) => n.to_string(),
        Value::String(s) => {
            if needs_quoting(s) {
                Value::String(s.clone()).to_string()
            } else {
                s.clone()
            }
        }
        other => other.to_string(),
    }
}

fn decode_scalar(token: &str) -> Value {
    if token.starts_with('"') {
        if let Ok(v) = serde_json::from_str::<Value>(token) {
            return v;
        }
    }
    match token {
        "null" => return Value::Null,
        "true" => return Value::Bool(true),
        "false" => return Value::Bool(false),
        _ => {}
    }
    if LOOKS_NUMERIC.is_match(token) {
        if let Ok(v) = serde_json::from_str::<Value>(token) {
            return v;
        }
    }
    Value::String(token.to_string())
}

fn escape_key(key: &str) -> String {
    key.replace('\\', "\\\\").replace('.', "\\.")
}

fn unescape_key(key: &str) -> String {
    key.replace("\\.", ".").replace("\\\\", "\\")
}

fn split_path(path: &str) -> Vec<String> {
    let mut parts = Vec::new();
    let mut buf = String::new();
    let mut chars = path.chars().peekable();
    while let Some(c) = chars.next() {
        if c == '\\' {
            if let Some(n) = chars.next() {
                buf.push(c);
                buf.push(n);
            }
            continue;
        }
        if c == '.' {
            parts.push(unescape_key(&buf));
            buf.clear();
            continue;
        }
        buf.push(c);
    }
    parts.push(unescape_key(&buf));
    parts
}

fn flatten(node: &Value, prefix: &str, sink: &mut Vec<(String, String)>) {
    match node {
        Value::Object(map) => {
            if map.is_empty() {
                sink.push((prefix.to_string(), "{}".to_string()));
                return;
            }
            for (k, v) in map {
                let child = escape_key(k);
                let path = if prefix.is_empty() { child } else { format!("{}.{}", prefix, child) };
                flatten(v, &path, sink);
            }
        }
        Value::Array(items) => {
            if items.is_empty() {
                sink.push((prefix.to_string(), "[]".to_string()));
                return;
            }
            let all_scalar = items.iter().all(|v| !matches!(v, Value::Object(_) | Value::Array(_)));
            if all_scalar {
                let joined = items.iter().map(encode_scalar).collect::<Vec<_>>().join(", ");
                sink.push((prefix.to_string(), joined));
                return;
            }
            for (i, v) in items.iter().enumerate() {
                let path = if prefix.is_empty() { i.to_string() } else { format!("{}.{}", prefix, i) };
                flatten(v, &path, sink);
            }
        }
        scalar => sink.push((prefix.to_string(), encode_scalar(scalar))),
    }
}

/// Flatten nested JSON into `path value` lines with no brackets and minimal quoting.
pub fn json_to_minimal_yaml(data: &Value) -> String {
    let mut sink = Vec::new();
    flatten(data, "", &mut sink);
    sink.iter()
        .map(|(p, v)| format!("{} {}", p, v).trim_end().to_string())
        .collect::<Vec<_>>()
        .join("\n")
}

fn split_top_level(text: &str) -> Vec<String> {
    let mut out = Vec::new();
    let mut buf = String::new();
    let mut in_str = false;
    let mut chars = text.chars().peekable();
    while let Some(c) = chars.next() {
        if in_str {
            if c == '\\' {
                buf.push(c);
                if let Some(n) = chars.next() {
                    buf.push(n);
                }
                continue;
            }
            if c == '"' {
                in_str = false;
            }
            buf.push(c);
        } else if c == '"' {
            in_str = true;
            buf.push(c);
        } else if c == ',' {
            out.push(buf.clone());
            buf.clear();
        } else {
            buf.push(c);
        }
    }
    out.push(buf);
    out
}

fn relist(node: Value) -> Value {
    match node {
        Value::Object(map) => {
            let converted: Map<String, Value> =
                map.into_iter().map(|(k, v)| (k, relist(v))).collect();
            let all_digits = !converted.is_empty() && converted.keys().all(|k| k.chars().all(|c| c.is_ascii_digit()));
            if all_digits {
                let mut keys: Vec<&String> = converted.keys().collect();
                keys.sort_by_key(|k| k.parse::<usize>().unwrap_or(usize::MAX));
                let expected: Vec<String> = (0..converted.len()).map(|i| i.to_string()).collect();
                if keys.iter().map(|k| k.as_str()).collect::<Vec<_>>()
                    == expected.iter().map(|k| k.as_str()).collect::<Vec<_>>()
                {
                    return Value::Array(expected.iter().map(|k| converted[k].clone()).collect());
                }
            }
            Value::Object(converted)
        }
        other => other,
    }
}

/// Inverse of [`json_to_minimal_yaml`]; proves the transform is lossless.
pub fn minimal_yaml_to_json(text: &str) -> Value {
    let mut root = Value::Object(Map::new());

    for line in text.lines() {
        if line.trim().is_empty() {
            continue;
        }
        let (path, raw_value) = match line.find(' ') {
            Some(idx) => (&line[..idx], line[idx + 1..].trim()),
            None => (line, ""),
        };
        let parts = split_path(path);

        let value = if raw_value == "{}" {
            Value::Object(Map::new())
        } else if raw_value == "[]" {
            Value::Array(Vec::new())
        } else if raw_value.contains(',') && !raw_value.starts_with('"') {
            Value::Array(split_top_level(raw_value).iter().map(|v| decode_scalar(v.trim())).collect())
        } else {
            decode_scalar(raw_value)
        };

        let mut cursor = &mut root;
        for (i, part) in parts.iter().enumerate() {
            let last = i == parts.len() - 1;
            let obj = cursor.as_object_mut().expect("object cursor");
            if last {
                obj.insert(part.clone(), value.clone());
            } else {
                cursor = obj.entry(part.clone()).or_insert_with(|| Value::Object(Map::new()));
            }
        }
    }

    relist(root)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn round_trip(v: Value) {
        let flat = json_to_minimal_yaml(&v);
        assert_eq!(minimal_yaml_to_json(&flat), v, "flat was:\n{}", flat);
    }

    #[test]
    fn simple() {
        round_trip(json!({"a": 1}));
    }

    #[test]
    fn deep_nesting() {
        round_trip(json!({"a": {"b": {"c": {"d": "deep"}}}}));
    }

    #[test]
    fn scalar_arrays() {
        round_trip(json!({"list": [1, 2, 3]}));
    }

    #[test]
    fn object_arrays() {
        round_trip(json!({"objs": [{"x": 1}, {"x": 2}]}));
    }

    #[test]
    fn empties() {
        round_trip(json!({"d": {}, "l": []}));
    }

    #[test]
    fn ambiguous_strings_stay_quoted() {
        let v = json!({"num_str": "123", "bool_str": "true", "version": "15.4"});
        let flat = json_to_minimal_yaml(&v);
        assert!(flat.contains("num_str \"123\""));
        round_trip(v);
    }

    #[test]
    fn unambiguous_strings_lose_quotes() {
        let flat = json_to_minimal_yaml(&json!({"name": "payment-api"}));
        assert_eq!(flat, "name payment-api");
    }

    #[test]
    fn no_bracket_noise() {
        let flat = json_to_minimal_yaml(&json!({"a": {"b": [1, 2]}}));
        assert!(!flat.contains('{'));
        assert!(!flat.contains('['));
    }

    #[test]
    fn keys_with_dots() {
        round_trip(json!({"key.with.dots": "value"}));
    }

    #[test]
    fn unicode_and_padding() {
        round_trip(json!({"u": "café → ☃", "p": "  padded  ", "e": ""}));
    }

    #[test]
    fn smaller_than_pretty_json() {
        let v = json!({"service": {"name": "api", "port": 8080, "tags": ["a", "b"]}});
        let flat = json_to_minimal_yaml(&v);
        let pretty = serde_json::to_string_pretty(&v).unwrap();
        assert!(flat.len() < pretty.len());
    }
}
